Add a navigation mode for message scroll

cobwebsonsale opened the pull request:
- Disable scroll to bottom on new messages if user
  scrolls up to read messages

cobwebsonsale committed:
Add a navigation mode for message scroll

- Disable scroll to bottom on new messages if user
  scrolls up to read messages

cobwebsonsale commented:
here's how the behaviour looks like:
![nav_mode](https://user-images.githubusercontent.com/1848127/96360855-68deab80-113e-11eb-81ce-1e45c6c67d36.gif)

olavfosse commented:
Will be reviewing this now.

olavfosse commented on front/src/components/Messages.js:
```diff
@@ -55,14 +55,21 @@ const MessagesDiv = styled.div`
 
 
 const Messages = (props) => {
+  const [isNavigationMode, setNavigationMode] = useState()
```
`const [isNavigationMode, setIsNavigationMode] = useState(false)`
is better

cobwebsonsale replied:
Can I push another commit for this edit or should I amend my last commit? How do you prefer things?

olavfosse replied:
push another. It's best to avoid changing public history.

cobwebsonsale replied:
Cool. This is done.

olavfosse commented:
Exiting navigation-mode, by scrolling all the way down, did not work on my box because `element.scrollTop + element.offSetHeight` and `element.scrollHeight` are slightly different even though I've scrolled all the way down I can. I'll fix this by changing the condition to only require that they are very close to each other. I'm thinking `3px` should be sufficient.

---
```
$ uname -a
Linux Shiina 5.4.0-52-generic #57-Ubuntu SMP Thu Oct 15 10:57:00 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

```
$ chromium --version
Chromium 86.0.4240.111 snap
```

olavfosse commented on front/src/components/Messages.js:
```diff
@@ -55,14 +55,21 @@ const MessagesDiv = styled.div`
 
+  const handleScroll = (scrollEvent) => {
+      var element = scrollEvent.currentTarget
+      var isBottom = element.scrollTop + element.offsetHeight === element.scrollHeight
```
avoid `var`. Use `const` for variables that you don't mutate and `let` for those that you do.

olavfosse commented on front/src/components/Messages.js:
```diff
@@ -55,14 +55,21 @@ const MessagesDiv = styled.div`
+  const handleScroll = (scrollEvent) => {
+      var element = scrollEvent.currentTarget
+      var isBottom = element.scrollTop + element.offsetHeight === element.scrollHeight
+      isBottom ? setNavigationMode(false) : setNavigationMode(true)
```
This is logically equivalent to
`setNavigationMode(isBottom)`

cobwebsonsale replied:
oh you're right. let me fix this.

olavfosse replied:
Sorry, it should of course be `setNavigationMode(!isBottom)`.

cobwebsonsale replied:
Done. Changed `isBottom` to `!isBottom` though.

cobwebsonsale committed:
Initialise isNavigationMode to false

cobwebsonsale commented:
>  I'll fix this by changing the condition to only require that they are very close to each other. I'm thinking 3px should be sufficient.

Makes sense. Stackoverflow had similar suggestions around testing the difference of two values to be less than 1 unit. (unsure what the unit was)

cobwebsonsale committed:
Turn off navigationMode on reaching bottom with an error margin

cobwebsonsale commented:
I've added the error margin for reaching bottom. Can you please check if it's enough when you scroll?

olavfosse commented:
Works great here!

olavfosse commented:
I am just adding some comments and small refinements then I will merge.

cobwebsonsale commented:
Safari/Mac
``` 
$ uname -a
Darwin 19.6.0 Darwin Kernel Version 19.6.0: Thu Jun 18 20:49:00 PDT 2020; root:xnu-6153.141.1~1/RELEASE_X86_64 x86_64
```
```
Safari
Version 14.0 (15610.1.28.1.9, 15610)
```

The browser actually scrolls _beyond_ the height, where the difference becomes negative.

olavfosse committed:
Refine #49

This commit just contains some new comments and small changes to
make it more readable.

olavfosse commented:
Merging now. Thanks!
